### PR TITLE
[MRG] Sampling iterables

### PR DESCRIPTION
**Pull request by tommyod**
Work in progress. Addresses #344.

## TODO
- [x] Review literature.
- [ ] Sampling with/without replacement. 
- [x] Sampling with/without weights.
- [x] Perform empirical analysis of running time.
- [x] Decide on function signature.

Decided not to add sampling with replacement in this PR. Three reasons: (1) with a large number of elements, sampling with replacement is not that interesting, (2) there's no general algorithm, so we would have to split up into four cases: weighted/unweighted and with/without replacement and (3) I couldn't find a paper for the weighted case with replacements. Could be generalized from the other papers, but I don't want to spend time trying to do the math right now. Out of scope for this PR.

### CODE REVIEW

**Comment by tommyod** [comment]
Sampling could either **return a list** after consuming the entire iterable, e.g.,
```python
samples = sample(range(5), k=3)
print(samples) # [3, 1, 4]
```
or it could **yield the reservoir** (current list of samples) after consuming each item, e.g.,
```python
for samples in sample(range(5), k=3):
    print(samples)

# [0]
# [0, 1]
# [0, 1, 2]
# [3, 1, 2]
# [3, 1, 4]
```
-------------
Thoughts? The first option might be more intuitive (and more in line with current APIs in e.g. `random`), but it consumes the entire iterable before returning. The second approach strikes me as more flexible, but perhaps less intuitive.

**Comment by bbayles** [comment]
>  I favor the second approach.
Yes, me too.

File: MANIFEST.in [hunk 1]
```diff
@@ -7,5 +7,4 @@ include docs/conf.py
 include docs/_static/*
 include fabfile.py
 include tox.ini
-include more_itertools/py.typed
```
Comment (bbayles) [inline MANIFEST.in]: Hmm, looks like some merge issues caused this to be added to the diff. Could you remove it?

File: docs/api.rst [hunk 2]
```diff
@@ -235,7 +236,7 @@ Others
 .. autofunction:: always_reversible
 .. autofunction:: side_effect
 .. autofunction:: iterate
-.. autofunction:: difference(iterable, func=operator.sub)
+.. autofunction:: difference(iterable, func=operator.sub, *, initial=None)
```
Comment (bbayles) [inline docs/api.rst]: I don't think this should be in the diff either.

File: docs/conf.py [hunk 3]
```diff
@@ -49,7 +49,7 @@
 # built documents.
 #
 # The short X.Y version.
-version = '7.2.0'
+version = '8.0.2'
```
Comment (bbayles) [inline docs/conf.py]: Another change to remove from the diff.

File: docs/versions.rst [hunk 4]
```diff
@@ -5,6 +5,45 @@ Version History
 .. automodule:: more_itertools
    :noindex:
 
+8.0.2
```
Comment (bbayles) [inline docs/versions.rst]: This file shouldn't be in the diff.

File: more_itertools/__init__.py [hunk 5]
```diff
@@ -1,2 +1,2 @@
-from more_itertools.more import *  # noqa
-from more_itertools.recipes import *  # noqa
+from .more import *  # noqa
```
Comment (bbayles) [inline more_itertools/__init__.py]: This file shouldn't be in the diff.

File: more_itertools/__init__.pyi [hunk 6]
```diff
@@ -0,0 +1,2 @@
+from .more import *
```
Comment (bbayles) [inline more_itertools/__init__.pyi]: This file shouldn't be in the diff.

File: more_itertools/more.py [hunk 7]
```diff
@@ -1,24 +1,14 @@
...
-    groupby,
-    islice,
-    repeat,
-    starmap,
-    takewhile,
-    tee,
-    zip_longest,
-)
-from operator import itemgetter, lt, gt, sub
+from itertools import (chain, compress, count, cycle, dropwhile, groupby,
```
Comment (bbayles) [inline more_itertools/more.py]: Could you avoid this re-formatting?

File: setup.py [hunk 8]
```diff
@@ -22,18 +22,18 @@ def get_long_description():
 
 setup(
     name='more-itertools',
-    version='7.2.0',
+    version='8.0.2',
```
Comment (bbayles) [inline setup.py]: This file shouldn't be in the diff.

**Comment by tommyod** [comment]
- I messed up the PR, but fixed it and force pushed now.

**Comment by bbayles** [comment]
Thanks for fixing the merge issues - much cleaner now.

Re: function signature, I'd get rid of the `k=1` and just let that be a positional argument?

Re: splitting the weighted and unweighted - this seems fine.

Re: Tests, I would suggest checking out the method [here](https://stackoverflow.com/questions/47164359/how-to-use-a-random-seed-value-in-order-to-unittest-a-prng-in-python).

I'm having second thoughts about "yield the reservoir." It appealed at first, but playing with how it works, I think the common use case would be: "just give me the last one." What do you think?

**Comment by tommyod** [comment]
Thanks for the feedback, will address these comments :+1:

> I'm having second thoughts about "yield the reservoir." It appealed at first, but playing with how it works, I think the common use case would be: "just give me the last one." What do you think?

I agree completely that the most common use case would be likely "give me the last one," which could be written as `last(sample(iterable, k))` if we "yield". If we consume the entire thing and "return", the drawback is that the user loses control and will have to consume the entire iterable. Unless the user does something like `sample(islice(iterable, n), k)` or `sample(time_limited(seconds, iterable)), k)`.

Three options:
1. We "yield" as sketched in the code now, but include `last(sample(iterable, k))` in the docstring to help users along.
2. We "return" after consuming the iterable. This is perhaps more intuitive, but the way I see it the user loses some control. The "yield" alternative reduces to the "return" alternative, but not the other way around.
3. We include a keyword `sample(iterable, k, yield_reservoir=False)`. This will allow for both use cases, but to me a function that's either a generator function or a normal function depending on a keyword feels a bit wrong.

I still slightly favor alternative (1) above since it's the most general one. However, I'll happily implement any of the three alternatives. I suppose there's a tradeoff between generality and what a user might expect here. Thoughts @bbayles ?

**Comment by bbayles** [comment]
Apologies for backtracking, but I think we should just return the final list and not yield intermediate ones. If the user wants to get results from something other than the full iterable, they can pass in an `islice`-ed version.

**Comment by tommyod** [comment]
No worries! Will do. Might take some time due to christmas, but I'll finish up the PR and ask for another review soon-ish.

**Comment by tommyod** [comment]
Ready for a review round. Some comments below:

## Speed

I'd say the algorithms are competitive in terms of speed for most inputs. They are asymptotically optimal, and the most complex algorithms from the papers are implemented (using so-called skip distributions). Their primary virtue is of course that they are memory efficient and work on iterables of unknown length. Notice below that `random.choices` actually samples with replacement. I just compared to this algorithm to get a feeling for speed.

```
N=10000000, k=1000
======== UNWEIGHTED SAMPLING ========
Unweighted sampling (_sample_unweighted) ran in: 0.39156
Unweighted sampling (random.sample) ran in: 0.0815
Unweighted sampling (np.random.choice) ran in: 0.87555
======== WEIGHTED SAMPLING ========
Weighted sampling (_sample_weighted) ran in: 0.71794
Weighted sampling (random.choices) ran in: 0.45808
Weighted sampling (np.random.choice) ran in: 1.24391

N=10000000, k=100000
======== UNWEIGHTED SAMPLING ========
Unweighted sampling (_sample_unweighted) ran in: 0.92425
Unweighted sampling (random.sample) ran in: 0.1321
Unweighted sampling (np.random.choice) ran in: 0.85672
======== WEIGHTED SAMPLING ========
Weighted sampling (_sample_weighted) ran in: 1.88625
Weighted sampling (random.choices) ran in: 0.65419
Weighted sampling (np.random.choice) ran in: 1.35635

N=10000000, k=1000000
======== UNWEIGHTED SAMPLING ========
Unweighted sampling (_sample_unweighted) ran in: 2.99556
Unweighted sampling (random.sample) ran in: 0.75991
Unweighted sampling (np.random.choice) ran in: 0.85322
======== WEIGHTED SAMPLING ========
Weighted sampling (_sample_weighted) ran in: 14.58644
Weighted sampling (random.choices) ran in: 2.45623
Weighted sampling (np.random.choice) ran in: 2.20805

```

## Numerical stability

Weights in the range 1E-100 to 1E100 have been tested and give the same result. In other words, the algorithm is invariant with respect to scaling such as `new_weights = [weight * scale for weight in weights]` and I've tested with this `scale = 10**100` and `scale = 10**-100`. I implemented the weighted sampling algorithm in log-space to increase numerical stability. Works well.

## Bias

The plot below shows the difference between means in many samples, when permuting the order of the iterable (of course the sampling algorithms should be invariant w.r.t. such permutations). There doesn't seem to be any bias. This is a sign that I've managed to implement the algorithm correctly. I had a bias of ~ 0.4 earlier here, found the bug and now the bias is 0.0032 in the plot.

image

## Review

Please look over and comment on anything at all. I kept some notation from one of the papers, but in general I have tried to use decent variable names.

Thanks.

File: more_itertools/more.py [hunk 9]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
...
+
+    # Generate random number that's the largest in a sample of k U(0,1) numbers
+    # Largest order statistic: https://en.wikipedia.org/wiki/Order_statistic
+    W = exp(log(random()) / k)
+
+    # The number of elements to skip before changing the reservoir is a random
+    # number with a geometric distribution. Sample it using random() and logs.
+    next_index = k + floor(log(random()) / log(1 - W)) + 1
+
+    for index, element in enumerate(iterable, k + 1):
```
Comment (bbayles) [inline more_itertools/more.py]: Why `k + 1` here? We're at position k, yes?

File: more_itertools/more.py [hunk 10]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
             yield function(item)
         except exceptions:
             pass
+
+
+def _sample_unweighted(iterable, k):
+    # Implementation of "Algorithm L" from the 1994 paper by Kim-Hung Li:
```
Comment (bbayles) [inline more_itertools/more.py]: For my own notes, the algorithm in the paper is:

image

File: more_itertools/more.py [hunk 11]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
...
+    # "Reservoir-Sampling Algorithms of Time Complexity O(n(1+log(N/n)))".
+
+    # Fill up the reservoir (collection of samples) with the first `k` samples
+    reservoir = list(islice(iterable, k))
```
Comment (bbayles) [inline more_itertools/more.py]:
```suggestion
    reservoir = take(k, iterable)
```

File: more_itertools/more.py [hunk 12]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
...
+def _sample_weighted(iterable, k, weights):
+    # Implementation of "A-ExpJ" from the 2006 paper by Efraimidis et al. :
+    # "Weighted random sampling with a reservoir".
+
+    # Log-transform for numerical stability for weights that are small/large
+    weight_keys = (log(random()) / weight for weight in weights)
+
+    # Fill up the reservoir (collection of samples) with the first `k`
+    # weight-keys and elements, then heapify the list.
+    reservoir = list(islice(zip(weight_keys, iterable), k))
```
Comment (bbayles) [inline more_itertools/more.py]:
```suggestion
    reservoir = take(k, zip(weight_keys, iterable))
```

File: tests/test_more.py [hunk 13]
```diff
@@ -2936,3 +2937,57 @@ def test_multiple(self):
...
+        data = list(range(100))
+        data_rev = list(reversed(data))
+
+        # Sample each data set 10 times
+        data_means = [mean(mi.sample(data, k=50)) for _ in range(10)]
+        data_rev_means = [mean(mi.sample(data_rev, k=50)) for _ in range(10)]
+
+        # The difference in the means should be low, i.e. little bias
+        difference_in_means = abs(mean(data_means) - mean(data_rev_means))
+        self.assertTrue(difference_in_means < 5)
```
Comment (bbayles) [inline tests/test_more.py]: Did you try the "use a fixed seed" [thing](https://stackoverflow.com/a/47164461)? I think I like that better than the arbitrary threshold here.

**Comment by bbayles** [comment]
Great work. A few questions and suggestions - let me know what you think.

File: more_itertools/more.py (hunk 9 above)
Comment (tommyod) [reply more_itertools/more.py]: Yes, adding 1 in that line and the one above can both be removed. :+1:

File: tests/test_more.py (hunk 13 above)
Comment (tommyod) [reply tests/test_more.py]: Addressing this in a comment below :+1:

**Comment by tommyod** [comment]
I agree with all your comments and have committed them.

Regarding **testing the functions**, I don't believe seeding the random number generator and testing "actual vs expected" provides much value here. It might even provide a false sense of security. Consider the following:

I could implement a complete junk algorithm for sampling (e.g. take k first elements, skip every remaining element) and a unit test with a fixed seed would pass. On the other hand, I could implement a more naive algorithm for sampling (e.g. algorithms without skip distributions from the papers) that's completely correct, but the unit test with a fixed seed would then fail since this new algorithm uses the random number generator differently.

I've added a test with a fixed random seed anyway (but I don't think it's very useful). I've also kept my property-based tests. I've highlighted in code comments that the thresholds are not arbitrary - they are set based on 10000 simulations. Though stochastic, these tests verify properties that *any* sampling algorithm should have. Unlike fixed seed unit tests, the property-based tests would immediately fail on a junk algorithm, and would pass for sampling algorithms that are correct regardless of implementation details. The false alarm rate should be less than 1/10000 - I hope that's OK. If not I can make it lower.

Hope the above makes sense. :+1:  Other have said it better than me, e.g. ["I don't think unit tests are the right tool for testing randomness."](https://softwareengineering.stackexchange.com/questions/147134/how-should-i-test-randomness/147142#147142).

File: more_itertools/more.py [hunk 14]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
...
+            reservoir[randrange(k)] = element
+            # The new W is the largest in a sample of k U(0, `old_W`) numbers
+            W *= exp(log(random()) / k)
+            next_index += floor(log(random()) / log(1 - W)) + 1
+
+    return reservoir
+
+
+def _sample_weighted(iterable, k, weights):
+    # Implementation of "A-ExpJ" from the 2006 paper by Efraimidis et al. :
```
Comment (bbayles) [inline more_itertools/more.py]: For my notes, the algorithm in the paper is:

image

File: more_itertools/more.py [hunk 15]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
...
+    smallest_weight_key, _ = reservoir[0]
+    weights_to_skip = log(random()) / smallest_weight_key
+
+    for weight, element in zip(weights, iterable):
+
+        if weight >= weights_to_skip:
+            # The notation here is consistent with the paper, but we store
+            # the weight-keys in log-space for better numerical stability.
+            smallest_weight_key, _ = reservoir[0]
+            t_w = exp(weight * smallest_weight_key)
```
Comment (bbayles) [inline more_itertools/more.py]: `smallest_weight_key` is `log(T_w)`, and we need `t_w = T_w^w_i`. `exp(log(T_w) * w)` is equivalent.

File: more_itertools/more.py [hunk 16]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
...
+            r_2 = uniform(t_w, 1)  # generate U(t_w, 1)
+            weight_key = log(r_2) / weight
```
Comment (bbayles) [inline more_itertools/more.py]: This is step 9 in the algorithm

File: more_itertools/more.py [hunk 17]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
...
+            heapreplace(reservoir, (weight_key, element))
+            smallest_weight_key, _ = reservoir[0]
```
Comment (bbayles) [inline more_itertools/more.py]: This is step 10 in the algorithm

File: more_itertools/more.py [hunk 18]
```diff
@@ -2689,3 +2691,93 @@ def map_except(function, iterable, *exceptions):
...
+    for weight, element in zip(weights, iterable):
+
+        if weight >= weights_to_skip:
+            # The notation here is consistent with the paper, but we store
+            # the weight-keys in log-space for better numerical stability.
+            smallest_weight_key, _ = reservoir[0]
+            t_w = exp(weight * smallest_weight_key)
+            r_2 = uniform(t_w, 1)  # generate U(t_w, 1)
+            weight_key = log(r_2) / weight
+            heapreplace(reservoir, (weight_key, element))
```
Comment (bbayles) [inline more_itertools/more.py]: This is step 8 in the algorithm

**Comment by bbayles** [comment]
Re: testing with a fixed seed, here's what I was suggesting: given a handful of items and a known stream of values from the RNG, we could work through the algorithms with pencil and paper and produce a good result. Then we could run the code with those same parameters and ensure that it produces the same result.

Make sense? If the "expected" values of the tests that are currently written aren't from that method, I agree they're not useful. But if they were worked out in advance, then they're good.

**Comment by tommyod** [comment]
Not entirely sure if I understand. *If* I understand correctly: Creating such a test it would force me, as the author, to step through the algorithm like a debugger. This is a useful exercise for me, but I don't see how it provides value as a test in a test suite. Values from the current test are ad-hoc, so we agree that the test as it stands is not very valuable.

Anyway, I've changed from WIP to MRG and consider this PR to be done unless you have any further comments. You're of course very welcome to make changes to the branch before merging if you want to - whether that's adding a test or changing the comments after reading through the algorithm.

Your comments regarding the algorithm are correct :+1: . I'm fairly confident in the implementation: I've implemented it with care, it matches the numpy implementation and naive implementations, the properties are obeyed (e.g. scaling and permutations, as in the stochastic tests), and you have looked over it too.

**Comment by bbayles** [comment]
Thanks for the PR; I'll update before the next release.

**Comment by tommyod** [comment]
Great! Thanks for reviewing and merging @bbayles :+1:

**Pull request merged**